collector/receiver/otelarrowreceiver/internal/arrow/arrow.go

```diff
@@ -350,48 +366,192 @@ func (r *Receiver) anyStream(serverStream anyStreamServer, method string) (retEr
+		}
+	}
+	// bounded queue to memory limit based on incoming uncompressed request size and waiters.
+	// Acquire will fail immediately if there are too many waiters,
+	// or will otherwise block until timeout or enough memory becomes available.
+	err = r.boundedQueue.Acquire(ctx, int64(prevAcquiredBytes))
+	if err != nil {
+		return fmt.Errorf("breaking stream: %v", err)
+	}
+
+	resp := batchResp{
+		id:             req.GetBatchId(),
+		bytesToRelease: int64(prevAcquiredBytes),
+	}
+
+	var authErr error
+	if r.authServer != nil {
+		var newCtx context.Context
+		if newCtx, err = r.authServer.Authenticate(thisCtx, authHdrs); err != nil {
+			authErr = err
+		} else {
+			thisCtx = newCtx
+		}
+	}
+
+	// processAndConsume will process and send an error to the sender loop
+	go func() {
+		defer r.inFlightWG.Done() // done processing
+		err = r.processAndConsume(thisCtx, method, ac, req, serverStream, authErr, resp, sizeHeaderFound)
+		resp.err = err
+		pendingCh <- resp
+	}()
+
+	return nil
+}
+
+func (r *Receiver) srvReceiveLoop(ctx context.Context, serverStream anyStreamServer, pendingCh chan<- batchResp, method string, ac arrowRecord.ConsumerAPI) (retErr error) {
+	hrcv := newHeaderReceiver(ctx, r.authServer, r.gsettings.IncludeMetadata)
+	for {
+		select {
+		case <-ctx.Done():
+			return status.Error(codes.Canceled, "server stream shutdown")
+		default:
+			err := r.recvOne(ctx, serverStream, hrcv, pendingCh, method, ac)
+			if err != nil {
+				return err
 			}
-			return err
 		}
+	}
+}
 
-		// Check for optional headers and set the incoming context.
-		thisCtx, authHdrs, err := hrcv.combineHeaders(streamCtx, req.GetHeaders())
-		if err != nil {
-			// Failing to parse the incoming headers breaks the stream.
-			r.telemetry.Logger.Error("arrow metadata error", zap.Error(err))
-			return err
+func (r *Receiver) sendOne(serverStream anyStreamServer, resp batchResp) error {
+	// Note: Statuses can be batched, but we do not take
+	// advantage of this feature.
+	status := &arrowpb.BatchStatus{
+		BatchId: resp.id,
+	}
+	if resp.err == nil {
+		status.StatusCode = arrowpb.StatusCode_OK
+	} else {
+		status.StatusMessage = resp.err.Error()
+		switch {
+		case errors.Is(resp.err, arrowRecord.ErrConsumerMemoryLimit):
+			r.telemetry.Logger.Error("arrow resource exhausted", zap.Error(resp.err))
+			status.StatusCode = arrowpb.StatusCode_RESOURCE_EXHAUSTED
+		case consumererror.IsPermanent(resp.err):
+			r.telemetry.Logger.Error("arrow data error", zap.Error(resp.err))
+			status.StatusCode = arrowpb.StatusCode_INVALID_ARGUMENT
+		default:
+			r.telemetry.Logger.Debug("arrow consumer error", zap.Error(resp.err))
+			status.StatusCode = arrowpb.StatusCode_UNAVAILABLE
 		}
+	}
 
-		var authErr error
-		if r.authServer != nil {
-			var newCtx context.Context
-			if newCtx, err = r.authServer.Authenticate(thisCtx, authHdrs); err != nil {
-				authErr = err
-			} else {
-				thisCtx = newCtx
+	err := serverStream.Send(status)
```

Comment:
Is there a default timeout configuration in place?

Reply:
Hmm I thought we relied on keepalive settings for stream operation timeouts. Do we want a more explicit timeout added for stream Send and Recv operations? The grpc keepalive settings that can be configured https://github.com/open-telemetry/opentelemetry-collector/blob/main/config/configgrpc/configgrpc.go#L129, otherwise this goroutine will block until the status response is sent from the server.

Reply:
I hvaen't thought much about what happens if the Send() blocks here. It means the buffer is full, and we could have a shorter timeout. The caller will timeout on its own, but a hung stream will continue until the lifetime and that might not be great. Let's evaluate and file an issue if we see this as a problem.

Reply:
Sounds good.